autotest.pm

```diff
@@ -345,6 +350,44 @@ sub loadtestdir {
     }
 }
 
+sub parse_serial_output {
+    return -1 if !-e 'serial0';
```

Comment:
I think there is variable for this file name

Reply:
There's a serialfile instance variable in the backend but since is an instance variable is in a different scope.

Maybe i should make it class variable...

Reply:
Another idea would be to move this to the backend somehow. One way would be to put it in the backend base class then it can be overridden by backends which use a socket instead of a file or multiplex multiple serial streams through one file etc.

Reply:
@richiejp 

It's a bit tricky to move it to the backend. Please check now the current solution. :-)

Reply:
Sorry to be a pain, but this still does not look right to me. :-)

Maybe it would be better to send a list of regular expressions to the backend and the backend sends back any matches. Sending the regular expressions over JSON is annoying (it makes my previous comment about regexs irrelevant), but that is basically what we are doing with every wait_serial at least while using virtio_console.

Or you could just send all the serial output over JSON since the last request. The JSON serialiser will probably cope.

Sending the file name and file position to the front end only works for regular files which support seek. If it is a socket or something else then it won't work and we will need a workaround. This basically tightly couples the backend with autotest.

Reply:
@richiejp 

I'm not understanding your point... I'm sending the ouptut via JSON: https://github.com/os-autoinst/os-autoinst/pull/932/files#diff-07616258592e792ab88f484012b38370R483


$serial is the serial data, returned by the function read_serial:
https://github.com/os-autoinst/os-autoinst/pull/932/files#diff-e9f8b7109dd3e4e2e6ce710cd417d82fR711

Reply:
Ah, OK, my bad. The other thing which is confusing me is that you are sending the file position. It would be nice if the backend stores the file position and just sends all the text since the previous call to `read_serial(..., stream_id => 'kernel_log')` where `stream_id` allows us to specify which serial output we need without any implementation specific information. 

However implementing that correctly might be a much bigger job and it won't take long to refactor this feature once it is implemented, so this looks OK for now, but only because it is a very nice feature. :-)

Reply:
There's already a mechanism in place to store the current position of the serial, however since tests can read from serial as well, we would need to implement a mechanism to identify reads.

I also think that a function to tail the serial file is useful :-)